### PR TITLE
categoryDefs: typeDefs, Queries, Mutations

### LINKED ISSUE #47: Category TypeDefs

### PR DESCRIPTION
**Pull request by sarah-jensen**
Mutations for add/update/delete categories also update category field for associated tutorials

closes #47 #48

### CODE REVIEW

File: server/schemas/categoryDefs.js [hunk 1]
```diff
@@ -11,14 +10,89 @@ const categoryTypeDefs = gql`
...
+        if (!updatedCategory) {
+          throw new Error('Category not found');
+        }
+    
+        // Find tutorials associated with updated category
+        const tutorials = await Tutorial.find({ categories: updatedCategory._id });
+    
+        // Update the category in each associated tutorial
+        await Promise.all(
+          tutorials.map(function (tutorial) {
```
Comment (sendusyourbones) [inline server/schemas/categoryDefs.js]: Sorry to be annoying but I think this is a callback and so should stay an arrow function

File: server/schemas/categoryDefs.js [hunk 2]
```diff
@@ -11,14 +10,89 @@ const categoryTypeDefs = gql`
...
+        throw new Error('Category not found');
+      }
+
+      //find tutorials associated with the category
+      const tutorials = await Tutorial.find({ categories: 
+      categoryToDelete._id });
+
+      //remove the category from each associated tutorial
+      await Promise.all(
+        tutorials.map(function(tutorial) {
```
Comment (sendusyourbones) [inline server/schemas/categoryDefs.js]: Same comment about callback here.

File: server/schemas/categoryDefs.js [hunk 3]
```diff
@@ -11,14 +10,89 @@ const categoryTypeDefs = gql`
...
+          throw new Error('Category not found');
+        }
+    
+        // Find tutorials associated with updated category
+        const tutorials = await Tutorial.find({ categories: updatedCategory._id });
+    
+        // Update the category in each associated tutorial
+        await Promise.all(
+          tutorials.map(function (tutorial) {
+            tutorial.categories.forEach(function (categoryId, index) {
```
Comment (sendusyourbones) [inline server/schemas/categoryDefs.js]: Same comment about callback here.

File: server/schemas/categoryDefs.js (hunk 1 above)
Comment (sarah-jensen) [reply server/schemas/categoryDefs.js]: restored =>

File: server/schemas/categoryDefs.js (hunk 3 above)
Comment (sarah-jensen) [reply server/schemas/categoryDefs.js]: restored =>

File: server/schemas/categoryDefs.js (hunk 2 above)
Comment (sarah-jensen) [reply server/schemas/categoryDefs.js]: restored =>

File: server/schemas/categoryDefs.js [hunk 4]
```diff
@@ -11,15 +10,105 @@ const categoryTypeDefs = gql`
 
   type Query {
     categories: [Category]
+    category(_id: ID!): Category
+  }
+
+  type Mutation {
+    addCategory(tutorialId: ID!, category: String!): Category
```
Comment (sendusyourbones) [inline server/schemas/categoryDefs.js]: This needs to be updated to reflect the fact that `tutorialId` is no longer required

File: server/schemas/lesson.js
Comment (sendusyourbones) [inline server/schemas/lesson.js:1]: I think this file may have gotten added erroneously?

File: server/schemas/schema.js
Comment (sendusyourbones) [inline server/schemas/schema.js:1]: I think this file may have gotten added erroneously?

File: server/schemas/categoryDefs.js (hunk 4 above)
Comment (sarah-jensen) [reply server/schemas/categoryDefs.js]: removed `tutorialId: ID!`

File: server/schemas/schema.js
Comment (sarah-jensen) [reply server/schemas/schema.js:1]: re-removed!

File: server/schemas/lesson.js
Comment (sarah-jensen) [reply server/schemas/lesson.js:1]: removed

File: server/schemas/categoryDefs.js [hunk 5]
```diff
@@ -11,15 +10,105 @@ const categoryTypeDefs = gql`
...
+      categoryToDelete._id });
+
+      //remove the category from each associated tutorial
+      await Promise.all(
+        tutorials.map((tutorial) => {
+          tutorial.categories.pull(categoryToDelete._id);
+          return tutorial.save();
+        })
+      )
+      return await Category.findByIdAndDelete(_id);
```
Comment (rileydoneil) [inline server/schemas/categoryDefs.js]: Try catch ono any db such as findbyIdAndDelete

File: server/schemas/categoryDefs.js [hunk 6]
```diff
@@ -11,15 +10,105 @@ const categoryTypeDefs = gql`
...
+          { category }, 
+          { new: true }
+        );
+    
+        if (!updatedCategory) {
+          throw new Error('Category not found');
+        }
+    
+        // Find tutorials associated with updated category
+        const tutorials = await Tutorial.find({ categories: updatedCategory._id });
```
Comment (rileydoneil) [inline server/schemas/categoryDefs.js]: try catch on find

File: server/schemas/schema.js
Comment (rileydoneil) [inline server/schemas/schema.js:1]: I dont think we need a schema.js file anymore

**Review by rileydoneil — COMMENTED** [review]
Add the prettier styling so there are tabs in nested functions and there are two try catch blocks missing. Overall, great work Sarah!

File: server/schemas/schema.js
Comment (sarah-jensen) [reply server/schemas/schema.js:1]: yea, that's super weird - I have no memory of creating a file with that name!

File: server/schemas/categoryDefs.js (hunk 5 above)
Comment (sarah-jensen) [reply server/schemas/categoryDefs.js]: added if-statement to catch error on line 109

File: server/schemas/categoryDefs.js (hunk 6 above)
Comment (sarah-jensen) [reply server/schemas/categoryDefs.js]: added if-statement to catch error on line 71

**Review by rileydoneil — APPROVED** [review]
GOOD STUFF! Code looks clean and works perfectly

**Review by rileydoneil — APPROVED** [review]
Good Stuff!! Works good and looks good

**Pull request merged**